### PR TITLE
Fix Calypso browsing unloaded package artifacts from Iceberg

### PR DESCRIPTION
**Pull request by hernanmd**
Add a check in RGBehavior>>inheritsFrom: so unresolved superclasses gets out of an infinite recursion. See https://github.com/pharo-project/pharo/issues/9684 for details.
Implements #definitionString hooks to display the class definition String in the Calypso browser for non-loaded packages, using MCSnapshot information.
Add missing #needsSlotClassDefinition and #allowsShadowing in RGBehavior, which were causing red morph of death while browsing some methods.
Add #buildMicrodownUsing:withComment in RGBehavior, needed for Calypso when rendering the Comment tab.

### CODE REVIEW

**Comment by Ducasse** [comment]
THANKS!!!

**Comment by MarcusDenker** [comment]
known random failing tests on win:
-  [CI] System variable environment on windows does not use adequate encodings #11407

File: src/Ring-Core/RGBehavior.class.st [hunk 1]
```diff
@@ -222,6 +222,16 @@ RGBehavior >> asYetUnclassifiedProtocolName [
 	^ Protocol unclassified
 ]
 
+{ #category : #accessing }
+RGBehavior >> buildMicroDownUsing: aMicMicrodownTextualBuilder withComment: aString [
```
Comment (guillep) [inline src/Ring-Core/RGBehavior.class.st]: I wonder if this is ok here or this should be an extension method.

File: src/Ring-Core/RGBehavior.class.st [hunk 2]
```diff
@@ -486,17 +503,16 @@ RGBehavior >> includesSelector: aString [
...
 	| aSuperclass |
 	aSuperclass := self superclass.
-	[aSuperclass == nil]
-		whileFalse: 
-			[aSuperclass == aClass ifTrue: [^true].
-			aSuperclass := aSuperclass superclass].
-	^false
+	[ aSuperclass isNil ] whileFalse: [
+		aSuperclass == aClass ifTrue: [ ^ true ].
+		aSuperclass hasResolvedName ifFalse: [ ^ false ].
```
Comment (guillep) [inline src/Ring-Core/RGBehavior.class.st:504]: Yes, good catch. Could we have a test for this?

Comment (guillep) [reply src/Ring-Core/RGBehavior.class.st:504]: I mean, the case of iterating the hierarchy when a superclass is not resolved. That's an important ring case.

File: src/Ring-Core/RGPackage.class.st [hunk 3]
```diff
@@ -43,6 +43,16 @@ RGPackage >> addoptToParentStub [
 	^ self parent addPackage: self.
 ]
 
+{ #category : #accessing }
+RGPackage >> buildMicroDownUsing: aMicMicrodownTextualBuilder withComment: aString [
```
Comment (guillep) [inline src/Ring-Core/RGPackage.class.st]: Same as before. Should this be an extension method?

File: src/Tool-DependencyAnalyser-Tests/DAMessageSendAnalyzerTest.class.st [hunk 4]
```diff
@@ -13,6 +13,16 @@ DAMessageSendAnalyzerTest >> setUp [
...
 ]
 
+{ #category : #tests }
+DAMessageSendAnalyzerTest >> testRPackage [
+
+	self assert: (analyzer rPackage isKindOf: RPackage).
+	analyzer := DAMessageSendAnalyzer on: 'Unloaded-Dummy-Package'.
+	self
+		assert: (analyzer rPackage isKindOf: RPackage)
+		description: 'It test that rPackage still answer a RPackage even when the analyzer was instantiated with an unexisting or unloaded package'
```
Comment (guillep) [inline src/Tool-DependencyAnalyser-Tests/DAMessageSendAnalyzerTest.class.st]: I'm not fond of tests doing isKindOf, but ok.

This kind of tests prevent future extensibility. Anyways, if required the test can be removed in the future.

Comment (guillep) [reply src/Tool-DependencyAnalyser-Tests/DAMessageSendAnalyzerTest.class.st]: A stronger test should be to send a message to the `analyzer rPackage` asserting expected classes or methods in the package

```smalltalk
self assert: analyzer rPackage definedClasses equals: ...
```

That asserts that the returned object works correctly, catches the bug too, and leaves the door open to extensions :)

File: src/Tool-DependencyAnalyser/DAMessageSendAnalyzer.class.st [hunk 5]
```diff
@@ -130,7 +130,11 @@ DAMessageSendAnalyzer >> possibleDeadCode [
 
 { #category : #accessing }
 DAMessageSendAnalyzer >> rPackage [
-	^ packageAnalysis rPackageSet packages detect: [ :each | each packageName = self packageName ]
+	"Answer a <RPackage> matching the receiver's package name. If we are browsing an unloaded package, then answer a virtual RPackage so the message send analyzer could preserve its behavior with loaded packages in #manuallyResolvedDependencies"
```
Comment (guillep) [inline src/Tool-DependencyAnalyser/DAMessageSendAnalyzer.class.st]: good catch!

This is the kind of thing that a type inferencer should be able to catch super easily...

**Review by guillep — CHANGES_REQUESTED** [review]

File: src/Ring-Core/RGBehavior.class.st (hunk 1 above)
Comment (hernanmd) [reply src/Ring-Core/RGBehavior.class.st]: Ok, added as extension methods in the Microdown repository.

File: src/Ring-Core/RGPackage.class.st (hunk 3 above)
Comment (hernanmd) [reply src/Ring-Core/RGPackage.class.st]: Done

File: src/Tool-DependencyAnalyser-Tests/DAMessageSendAnalyzerTest.class.st (hunk 4 above)
Comment (guillep) [reply src/Tool-DependencyAnalyser-Tests/DAMessageSendAnalyzerTest.class.st]: I don't understand, the test is gone, is that on purpose?

File: src/Tool-DependencyAnalyser/DAMessageSendAnalyzer.class.st (hunk 5 above)
Comment (guillep) [reply src/Tool-DependencyAnalyser/DAMessageSendAnalyzer.class.st]: This fix is not in the PR anymre?

File: src/Tool-DependencyAnalyser-Tests/DAMessageSendAnalyzerTest.class.st (hunk 4 above)
Comment (hernanmd) [reply src/Tool-DependencyAnalyser-Tests/DAMessageSendAnalyzerTest.class.st]: I re-added the missing methods. 
Added as testRPackageIsAnsweredWithLoadedPackage

File: src/Tool-DependencyAnalyser/DAMessageSendAnalyzer.class.st (hunk 5 above)
Comment (hernanmd) [reply src/Tool-DependencyAnalyser/DAMessageSendAnalyzer.class.st]: I don't know what happened, but I retrieved the changes and re-commited again

**Review by MarcusDenker — APPROVED** [review]

**Pull request merged**